src/bloqade/pyqrack/target.py

```diff
@@ -64,49 +63,50 @@ def _get_interp(self, mt: ir.Method[Params, RetType]):
 
     def run(
         self,
-        mt: ir.Method[Params, RetType],
-        *args: Params.args,
-        **kwargs: Params.kwargs,
-    ) -> RetType:
+        mt: ir.Method[..., RetType],
+        shots: int = 1,
+        args: tuple[Any, ...] = (),
+        kwargs: dict[str, Any] = {},
+        return_iterator: bool = False,
+    ) -> RetType | list[RetType] | Iterator[RetType]:
         """Run the given kernel method on the PyQrack simulator.
 
         Args
             mt (Method):
                 The kernel method to run.
+            shots (int):
+                The number of shots to run the simulation for.
+                Defaults to 1.
+            args (tuple[Any, ...]):
+                Positional arguments to pass to the kernel method.
+                Defaults to ().
+            kwargs (dict[str, Any]):
+                Keyword arguments to pass to the kernel method.
+                Defaults to {}.
+            return_iterator (bool):
+                Whether to return an iterator that yields results for each shot.
+                Defaults to False. if False, a list of results is returned.
 
         Returns
-            The result of the kernel method, if any.
-
-        """
-        fold = Fold(mt.dialects)
-        fold(mt)
-        return self._get_interp(mt).run(mt, args, kwargs)
-
-    def multi_run(
-        self,
-        mt: ir.Method[Params, RetType],
-        _shots: int,
-        *args: Params.args,
-        **kwargs: Params.kwargs,
-    ) -> List[RetType]:
-        """Run the given kernel method on the PyQrack `_shots` times, caching analysis results.
-
-        Args
-            mt (Method):
-                The kernel method to run.
-            _shots (int):
-                The number of times to run the kernel method.
-
-        Returns
-            List of results of the kernel method, one for each shot.
+            RetType | list[RetType] | Iterator[RetType]:
+                The result of the simulation. If `return_iterator` is True,
+                an iterator that yields results for each shot is returned.
+                Otherwise, a list of results is returned if `shots > 1`, or
+                a single result is returned if `shots == 1`.
```

Comment:
I'm not sure if returning different data type here is a good idea. should we always return `list` or `Iterator[RetType]`? Maybe `Iterable[RetType]`?

Reply:
This is a typo It should be `Iterable`

Reply:
One option is we just always return the `Iterable` and then just let people capture those values into a list by default:

`result = list(device.run(...))`